(DRAFT) Combined draft branch for coverage work

Zalathar opened the pull request:
While waiting for #114843, juggling my patch queue is getting to be quite unwieldy, so I'm going to try putting all of my pending PRs (including some not filed yet) and some of my in-progress work in one linear branch.

---

r? @ghost

Zalathar commented:
@rustbot label +A-code-coverage +S-experimental

Zalathar commented:
@rustbot author

Zalathar commented:
Quick summary of some of the changes I have queued up:

- Add the new `coverage-map` test suite
- Get rid of expression renumbering in coverage codegen
- Simplify the `coverageinfo` query to just one pass
- Change `StatementKind::Coverage` to store a vector of code regions, instead of just 0-1
- Rename `Operand` to `CovTerm` so that it can represent things that aren't operands
- Extract a `GlobalFileTable` helper in `mapgen.rs`
- Minor improvements to `write_coverage_mapping`, and rename it to `encode_mappings_for_function`
- Change `FunctionCoverage` to store all of its mappings as `FxHashMap<Symbol, Vec<(CovTerm, CodeRegion)>>`
  - Mappings are grouped by filename (since this needs to happen anyway), then stored as a list of kind/region pairs
  - Currently just reusing `CovTerm` as the mapping kind (since we only emit `Code` mappings), but eventually this will need to be replaced with a `MappingKind` enum that can represent expansion/gap/skip/branch mappings as well
  - This is an important step on the road towards being able to store coverage mappings in a per-function table in MIR

bors commented:
:umbrella: The latest upstream changes (presumably #114843) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
@rustbot label -A-testsuite -T-bootstrap

bors commented:
:umbrella: The latest upstream changes (presumably #115671) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #114656) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #115894) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6220015006/job/16879199467) [(plain)](https://github.com/rust-lang/rust/commit/083ee9acadb1414fd6d32627b8836381a9700ccc/checks/16879199467/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Zalathar
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_3d3eb275-6352-48d9-a76b-37c421f1a2ed
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=9a80e7dcfd3d20d9dedca0930b13fb7d9593e135
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_3d3eb275-6352-48d9-a76b-37c421f1a2ed
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_3d3eb275-6352-48d9-a76b-37c421f1a2ed
GITHUB_TRIGGERING_ACTOR=Zalathar
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115428/merge
GITHUB_WORKFLOW_SHA=9a80e7dcfd3d20d9dedca0930b13fb7d9593e135
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
##[endgroup]
##[group]Testing stage2 compiletest suite=coverage-map mode=coverage-map (x86_64-unknown-linux-gnu)

running 43 tests
..........F.....F..........................
failures:

---- [coverage-map] tests/coverage-map/status-quo/closure_macro.rs stdout ----
diff of cov-map:
diff of cov-map:

7 - Code(Counter(0)) at (prev + 29, 1) to (start + 2, 2)
9 Function name: closure_macro::main
9 Function name: closure_macro::main
- Raw bytes (43): 0x[01, 01, 02, 01, 05, 05, 02, 07, 01, 21, 01, 01, 21, 02, 02, 09, 00, 0f, 05, 00, 12, 00, 13, 02, 00, 12, 00, 13, 05, 00, 54, 00, 55, 02, 02, 09, 02, 0b, 07, 03, 01, 00, 02]
+ Raw bytes (43): 0x[01, 01, 02, 01, 05, 05, 02, 07, 01, 21, 01, 01, 21, 02, 02, 09, 00, 0f, 02, 00, 12, 00, 13, 05, 00, 12, 00, 13, 05, 00, 54, 00, 55, 02, 02, 09, 02, 0b, 07, 03, 01, 00, 02]
12 - file 0 => global file 1
13 Number of expressions: 2


17 - Code(Counter(0)) at (prev + 33, 1) to (start + 1, 33)
18 - Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 0, 15)
19     = (c0 - c1)
- - Code(Counter(1)) at (prev + 0, 18) to (start + 0, 19)
21 - Code(Expression(0, Sub)) at (prev + 0, 18) to (start + 0, 19)
22     = (c0 - c1)
+ - Code(Counter(1)) at (prev + 0, 18) to (start + 0, 19)
23 - Code(Counter(1)) at (prev + 0, 84) to (start + 0, 85)
Build completed unsuccessfully in 0:12:37
24 - Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 2, 11)
25     = (c0 - c1)

The actual cov-map differed from the expected cov-map.
The actual cov-map differed from the expected cov-map.
Actual cov-map saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage-map/status-quo/closure_macro/closure_macro.cov-map
error: 1 errors occurred comparing coverage output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/coverage-dump" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage-map/status-quo/closure_macro/closure_macro.ll"
--- stdout -------------------------------
--- stdout -------------------------------
Function name: closure_macro::load_configuration_files
Raw bytes (9): 0x[01, 01, 00, 01, 01, 1d, 01, 02, 02]
- file 0 => global file 1
Number of expressions: 0
Number of file 0 mappings: 1
Number of file 0 mappings: 1
- Code(Counter(0)) at (prev + 29, 1) to (start + 2, 2)
Function name: closure_macro::main
Function name: closure_macro::main
Raw bytes (43): 0x[01, 01, 02, 01, 05, 05, 02, 07, 01, 21, 01, 01, 21, 02, 02, 09, 00, 0f, 02, 00, 12, 00, 13, 05, 00, 12, 00, 13, 05, 00, 54, 00, 55, 02, 02, 09, 02, 0b, 07, 03, 01, 00, 02]
- file 0 => global file 1
Number of expressions: 2
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
Number of file 0 mappings: 7
- Code(Counter(0)) at (prev + 33, 1) to (start + 1, 33)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 0, 15)
    = (c0 - c1)
- Code(Expression(0, Sub)) at (prev + 0, 18) to (start + 0, 19)
    = (c0 - c1)
- Code(Counter(1)) at (prev + 0, 18) to (start + 0, 19)
- Code(Counter(1)) at (prev + 0, 84) to (start + 0, 85)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 2, 11)
    = (c0 - c1)
- Code(Expression(1, Add)) at (prev + 3, 1) to (start + 0, 2)
    = (c1 + (c0 - c1))
Function name: closure_macro::main::{closure#0}
Function name: closure_macro::main::{closure#0}
Raw bytes (9): 0x[01, 01, 00, 01, 01, 23, 12, 00, 54]
- file 0 => global file 1
Number of expressions: 0
Number of file 0 mappings: 1
Number of file 0 mappings: 1
- Code(Counter(0)) at (prev + 35, 18) to (start + 0, 84)
stderr: none


---- [coverage-map] tests/coverage-map/status-quo/closure_macro_async.rs stdout ----
---- [coverage-map] tests/coverage-map/status-quo/closure_macro_async.rs stdout ----
diff of cov-map:

15 - Code(Counter(0)) at (prev + 33, 1) to (start + 0, 43)
17 Function name: closure_macro_async::test::{closure#0}
17 Function name: closure_macro_async::test::{closure#0}
- Raw bytes (43): 0x[01, 01, 02, 01, 05, 05, 02, 07, 01, 21, 2b, 01, 21, 02, 02, 09, 00, 0f, 05, 00, 12, 00, 13, 02, 00, 12, 00, 13, 05, 00, 54, 00, 55, 02, 02, 09, 02, 0b, 07, 03, 01, 00, 02]
+ Raw bytes (43): 0x[01, 01, 02, 01, 05, 05, 02, 07, 01, 21, 2b, 01, 21, 02, 02, 09, 00, 0f, 02, 00, 12, 00, 13, 05, 00, 12, 00, 13, 05, 00, 54, 00, 55, 02, 02, 09, 02, 0b, 07, 03, 01, 00, 02]
20 - file 0 => global file 1
21 Number of expressions: 2


25 - Code(Counter(0)) at (prev + 33, 43) to (start + 1, 33)
26 - Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 0, 15)
27     = (c0 - c1)
- - Code(Counter(1)) at (prev + 0, 18) to (start + 0, 19)
29 - Code(Expression(0, Sub)) at (prev + 0, 18) to (start + 0, 19)
30     = (c0 - c1)
+ - Code(Counter(1)) at (prev + 0, 18) to (start + 0, 19)
31 - Code(Counter(1)) at (prev + 0, 84) to (start + 0, 85)
32 - Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 2, 11)
33     = (c0 - c1)

The actual cov-map differed from the expected cov-map.
The actual cov-map differed from the expected cov-map.
Actual cov-map saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage-map/status-quo/closure_macro_async/closure_macro_async.cov-map
error: 1 errors occurred comparing coverage output.
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/coverage-dump" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage-map/status-quo/closure_macro_async/closure_macro_async.ll"
--- stdout -------------------------------
--- stdout -------------------------------
Function name: closure_macro_async::load_configuration_files
Raw bytes (9): 0x[01, 01, 00, 01, 01, 1d, 01, 02, 02]
- file 0 => global file 1
Number of expressions: 0
Number of file 0 mappings: 1
Number of file 0 mappings: 1
- Code(Counter(0)) at (prev + 29, 1) to (start + 2, 2)
Function name: closure_macro_async::test
Function name: closure_macro_async::test
Raw bytes (9): 0x[01, 01, 00, 01, 01, 21, 01, 00, 2b]
- file 0 => global file 1
Number of expressions: 0
Number of file 0 mappings: 1
Number of file 0 mappings: 1
- Code(Counter(0)) at (prev + 33, 1) to (start + 0, 43)
Function name: closure_macro_async::test::{closure#0}
Function name: closure_macro_async::test::{closure#0}
Raw bytes (43): 0x[01, 01, 02, 01, 05, 05, 02, 07, 01, 21, 2b, 01, 21, 02, 02, 09, 00, 0f, 02, 00, 12, 00, 13, 05, 00, 12, 00, 13, 05, 00, 54, 00, 55, 02, 02, 09, 02, 0b, 07, 03, 01, 00, 02]
- file 0 => global file 1
Number of expressions: 2
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
Number of file 0 mappings: 7
- Code(Counter(0)) at (prev + 33, 43) to (start + 1, 33)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 0, 15)
    = (c0 - c1)
- Code(Expression(0, Sub)) at (prev + 0, 18) to (start + 0, 19)
    = (c0 - c1)
- Code(Counter(1)) at (prev + 0, 18) to (start + 0, 19)
- Code(Counter(1)) at (prev + 0, 84) to (start + 0, 85)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 2, 11)
    = (c0 - c1)
- Code(Expression(1, Add)) at (prev + 3, 1) to (start + 0, 2)
    = (c1 + (c0 - c1))
Function name: closure_macro_async::test::{closure#0}::{closure#0}
Function name: closure_macro_async::test::{closure#0}::{closure#0}
Raw bytes (9): 0x[01, 01, 00, 01, 01, 23, 12, 00, 54]
- file 0 => global file 1
Number of expressions: 0
Number of file 0 mappings: 1
Number of file 0 mappings: 1
- Code(Counter(0)) at (prev + 35, 18) to (start + 0, 84)
stderr: none



```

</details>

Zalathar commented:
Ah, that failure probably comes from no longer explicitly sorting BCBs (since we now group them via a map). If so, the fix should be to sort coverage spans by BCB just before building the map.

bors commented:
:umbrella: The latest upstream changes (presumably #115952) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #115865) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #116006) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar committed:
coverage: Emit mappings for unused functions without generating stubs

Zalathar committed:
coverage: Move unused-function helpers closer to where they are used

Zalathar committed:
coverage: Repair instrumented functions that have lost all their counters

If a function has been instrumented for coverage, but MIR optimizations
subsequently remove all of its counter-increment statements, then we won't emit
LLVM counter-increment intrinsics. LLVM will think the function is not
instrumented, and it will disappear from coverage mappings and coverage
reports.

This new MIR pass detects when that has happened, and re-inserts a dummy
counter-increment statement so that LLVM knows to treat the function as
instrumented.

bors commented:
:umbrella: The latest upstream changes (presumably #116154) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar committed:
Allow MIR pass `UnreachableProp` when coverage is enabled

Zalathar committed:
coverage: Change query `codegened_and_inlined_items` to a plain function

This query has a name that sounds general-purpose, but in fact it has
coverage-specific semantics, and (fortunately) is only used by coverage code.

Because it is only ever called once, it doesn't need to be a query, and we can
change it to a regular function instead.

Zalathar committed:
coverage: Encode function mappings without re-sorting them

Zalathar committed:
coverage: Encapsulate local-to-global file mappings

Zalathar committed:
coverage: Build the global file table ahead of time

Zalathar committed:
coverage: Force `-Copt-level=2` in the unreachable-code test

This test is mainly for detecting problems triggered by MIR optimizations, but
the run-coverage tests don't enable optimization by default.

(The coverage-map tests do enable optimization by default, but I'm updating the
test anyway so that the two `unreachable.rs` files don't drift out of sync.)

Zalathar committed:
coverage: Split `FunctionCoverage` into distinct collector/finished phases

Zalathar commented:
Some of the changes I have queued up, that depend on #116046:

- Drastically simplify how the instrumentor injects coverage statements, removing a bunch of gunk that no longer serves a useful purpose
- Emit mappings for unused functions without having to synthesize stub implementations for them
- Repair instrumented functions that have lost all their counters to MIR optimizations
- Split `FunctionCoverage` into distinct collector/finished phases

Zalathar committed:
coverage: Fetch expressions and mappings separately

Zalathar committed:
coverage: Emit the filenames section before encoding per-function mappings

Most coverage metadata is encoded into two sections in the final executable.
The `__llvm_covmap` section mostly just contains a list of filenames, while the
`__llvm_covfun` section contains encoded coverage maps for each instrumented
function.

The catch is that each per-function record also needs to contain a hash of the
filenames list that it refers to. Historically this was handled by assembling
most of the per-function data into a temporary list, then assembling the
filenames buffer, then using the filenames hash to emit the per-function data,
and then finally emitting the filenames table itself.

However, now that we build the filenames table up-front (via a separate
traversal of the per-function data), we can hash and emit that part first, and
then emit each of the per-function records immediately after building. This
removes the awkwardness of having to temporarily store nearly-complete
per-function records.

Zalathar committed:
coverage: Consolidate creation of `__llvm_covmap` records

The code for creating these records was spread across multiple functions and
multiple files, so this change moves as much of it as possible into one place.

bors commented:
:umbrella: The latest upstream changes (presumably #116409) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #115214) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #116922) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
